### PR TITLE
feat: get product by ID

### LINKED ISSUE #86: [FEAT]: Single Products API
## Description

- Develop API endpoint that retrieves information about a single product. The response includes product details such as ID, name, description, price, and availability.

## Acceptance Criteria

- The endpoint `/api/v1/products/{product-id}` is accessible via a GET request.
- When a valid `product-id` is provided, the endpoint returns a 200 OK status.
- When an invalid `product-id` format is provided, the endpoint returns a 400 Bad Request status.
- On an invalid request, the response body contains an error message indicating the invalid product ID format.
- When a `product-id` that does not exist is provided, the endpoint returns a 404 Not Found status.
- For a product that does not exist, the response body contains an error message indicating that the product was not found.
- If an unexpected server error occurs, the endpoint returns a 500 Internal Server Error status.
- The response body of an error 500 response contains an error message indicating that an unexpected error occurred.
- Ensure that all unit and integration tests for the fetch single product endpoint are comprehensive, and verify the functionality correctly.

### Purpose

- This provides a backend service that allows fetching detailed information on selected/specific product

### Requirements

- [ ]  Implement an endpoint to view single product
- [ ]  Return detailed information on products
- [ ]  Handle situations where products don't exist

### Expected Outcome

- API endpoint that returns detailed information on specific product

### Tasks

- [ ]  Create an endpoint `/api/v1/products/{product-id}` to fetch details of a single product.
- [ ]  Implement server-side logic to handle successful and failed product retrieval requests.
- [ ]  Ensure all sensitive data is encrypted and secure.
- [ ]  Implement logging for errors and product retrieval attempts.
- [ ]  Write unit tests for the fetch single product endpoint.
- [ ]  Write integration tests for the fetch single product logic.
- [ ]  Perform security testing to ensure data protection and compliance.

## Database Schema

```
Table products {
  id int [pk, increment]
  name varchar(255) [not null]
  description text
  price decimal(10,2) [not null]
  available boolean [default: true]
  category varchar(255)
  created_at timestamp [default: `current_timestamp`]
  updated_at timestamp [default: `current_timestamp`]
}
```

### Success Response

- #### Code: 200 OK
    
- #### Content:
    

```
{
  "product": {
    "id": "string",
    "name": "string",
    "description": "string",
    "price": "number",
    "available": "boolean",
    "category": "string",
    "created_at": "string",
    "updated_at": "string"
  }
}
```

### Error Responses

- #### Code: 400 BAD REQUEST
    
- #### Content:
    

```
{
  "error": "Invalid product ID"
}
```

- #### Code: 404 NOT FOUND
    
- #### Content:
    

```
{
  "error": "Product not found"
}
```

- #### Code 422: UNPROCESSABLE CONTENT / VALIDATION ERROR
    

#### Content:

```
{
  "errors": {
    "field": "<field>",
    "message": "<message>"
  }
}
```

- #### Code: 500 INTERNAL SERVER ERROR
    
- #### Content:
    

```
{
  "error": "An unexpected error occurred"
}
```

### Fetch Single Product By ID

`curl -X GET "https:/<base-url>/api/v1/products/12345`

Reference: [Approved Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/181)

### PR DESCRIPTION
**Pull request by droffilc1**
# Description

**Closes #86**

# Changes proposed
- Implement an endpoint to view single product
- Return detailed information on products
- Handle situations where products don't exist

## What were you told to do?
- Create an endpoint /api/v1/products/{product-id} to fetch details of a single product.
- Implement server-side logic to handle successful and failed product retrieval requests.
 - Ensure all sensitive data is encrypted and secure.
 - Implement logging for errors and product retrieval attempts.
 - Write unit tests for the fetch single product endpoint.
 - Write integration tests for the fetch single product logic.
 - Perform security testing to ensure data protection and compliance.

## What did you do?
- Create an endpoint /api/v1/products/{product-id} to fetch details of a single product.
- Implement server-side logic to handle successful and failed product retrieval requests.
 - Ensure all sensitive data is encrypted and secure.
 - Implement logging for errors and product retrieval attempts.
 - Write unit tests for the fetch single product endpoint. 

# Check List (Check all the applicable boxes)

🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.
- [x] I am making a pull request against the **dev branch** (left side).
- [x] My commit messages styles matches our requested structure.
- [x] My code additions will fail neither code linting checks nor unit test.
- [x] I am only making changes to files I was requested to.

# Screenshots/Videos

### CODE REVIEW

**Comment by droffilc1** [comment]
@dmendie Am failing tests for whitespaces. I can't seem to see it in my editor. Kindly advise.

**Comment by chimaihueze** [comment]
@droffilc1 kindly run `dotnet format` and it'll format your spaces accordingly.

**Comment by droffilc1** [comment]
@chimaihueze I have done that but it seems to fail the check

**Comment by chimaihueze** [comment]
> @chimaihueze I have done that but it seems to fail the check

- The build error you're encountering now is due to a missing closing brace (}) in the ProductController.cs file at line 51. 

- Also, you have conflicts between different versions of the Microsoft.EntityFrameworkCore.Relational package. 

- Lastly, the title of the pull request, `feat: single product`, does not fully convey the purpose of this feature. For better clarity and to ensure that the title accurately reflects the added functionality, it should be renamed to something more descriptive. I suggest updating the title to "feat: get product by ID".

Fix these.

File: src/Hng.Application.Test/Features/Products/GetProductByIdQueryShould.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+{
+    public class GetProductByIdQueryShould
+    {
+        private readonly Mock<IRepository<Product>> _mockRepository;
+        private readonly Mock<IMapper> _mockMapper;
+
+        public GetProductByIdQueryShould()
+        {
+            _mockRepository = new Mock<IRepository<Product>>();
+            _mockMapper = new Mock<IMapper>();
```
Comment (NdubuisiJr) [inline src/Hng.Application.Test/Features/Products/GetProductByIdQueryShould.cs:21]: Don't mock IMapper. Take a look at other people's test

File: src/Hng.Application/Features/Products/Dtos/ProductDto.cs [hunk 2]
```diff
@@ -2,8 +2,13 @@ namespace Hng.Application.Features.Products.Dtos
 {
     public class ProductDto
     {
-        public Guid id { get; set; }
-        public string name { get; set; }
-        public string description { get; set; }
+        public Guid Id { get; set; }
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/Products/Dtos/ProductDto.cs:5]: Use `JsonPropertyName` attribute to turn all your request and response dtos to `snake_case`

File: src/Hng.Application/Features/Products/Handlers/GetProductByIdQueryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            _productRepository = productRepository;
+            _mapper = mapper;
+        }
+
+        public async Task<ProductDto> Handle(GetProductByIdQuery request, CancellationToken cancellationToken)
+        {
+            var product = await _productRepository.GetByIdAsync(request.Id);
+            if (product == null)
+            {
+                throw new KeyNotFoundException($"Product with ID {request.Id} not found.");
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/Products/Handlers/GetProductByIdQueryHandler.cs:27]: Don't throw exceptions. Return null.

File: src/Hng.Infrastructure/Repository/Interface/IRepository.cs [hunk 4]
```diff
@@ -18,5 +18,7 @@ public interface IRepository<T> where T : EntityBase
         Task<int> CountAsync();
         Task<int> CountAsync(Expression<Func<T, bool>> predicate);
         Task SaveChanges();
+
+        Task<T> GetByIdAsync(Guid id);
```
Comment (NdubuisiJr) [inline src/Hng.Infrastructure/Repository/Interface/IRepository.cs:22]: Take this none sense out of here. The Repository provides everything for you

File: src/Hng.Infrastructure/Repository/Repository.cs [hunk 5]
```diff
@@ -58,6 +58,12 @@ public async Task<T> GetAsync(Guid id)
             return await _context.Set<T>().FindAsync(id);
         }
 
+        public async Task<T> GetByIdAsync(Guid id)
+        {
```
Comment (NdubuisiJr) [inline src/Hng.Infrastructure/Repository/Repository.cs:62]: Take this out. Can't you see the method above this does the same thing?

File: src/Hng.Web/Controllers/ProductController.cs [hunk 6]
```diff
@@ -1,19 +1,52 @@
...
+using Hng.Application.Features.Products.Queries;
+using Hng.Application.Features.Products.Dtos;
+using MediatR;
 using Microsoft.AspNetCore.Mvc;
 
 namespace Hng.Web.Controllers
 {
     [ApiController]
-    [Route("api/v1/products")]
+    [Route("api/v1/[controller]")]
```
Comment (NdubuisiJr) [inline src/Hng.Web/Controllers/ProductController.cs:10]: Don't use conventional naming. write it yourself.

File: src/Hng.Web/Controllers/ProductController.cs [hunk 7]
```diff
@@ -1,19 +1,52 @@
...
         }
 
+        [HttpGet("{id}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public async Task<ActionResult<ProductDto>> GetProductById(Guid id)
+        {
+            try
```
Comment (NdubuisiJr) [inline src/Hng.Web/Controllers/ProductController.cs:29]: Remove all of this try catch

File: src/Hng.Web/Controllers/ProductController.cs [hunk 8]
```diff
@@ -1,19 +1,52 @@
-﻿using MediatR;
+using System;
+using Hng.Application.Features.Products.Queries;
+using Hng.Application.Features.Products.Dtos;
+using MediatR;
 using Microsoft.AspNetCore.Mvc;
 
 namespace Hng.Web.Controllers
 {
     [ApiController]
```
Comment (NdubuisiJr) [inline src/Hng.Web/Controllers/ProductController.cs:9]: Protect this controller.

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

File: src/Hng.Application/Features/Products/Dtos/ProductDto.cs [hunk 9]
```diff
@@ -6,4 +6,4 @@ public class ProductDto
         public string name { get; set; }
         public string description { get; set; }
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/Products/Dtos/ProductDto.cs:7]: Don't use camelCase for Properties. Use Normal PascalCasing, then apply `JsonPropertyName` attribute to but it in snake_case

File: src/src.sln [hunk 10]
```diff
@@ -0,0 +1,49 @@
+﻿
```
Comment (NdubuisiJr) [inline src/src.sln:1]: Revert all of the changes to this file

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

File: src/src.sln [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (NdubuisiJr) [inline src/src.sln:2]: Delete this file. There's no `src.sln`

File: src/Hng.Web/Controllers/ProductController.cs [hunk 12]
```diff
@@ -1,19 +1,39 @@
...
+        {
+            if (id == Guid.Empty)
+            {
+                return BadRequest(new { error = "Invalid product ID" });
+            }
+
+            var query = new GetProductByIdQuery(id);
+            var product = await _mediator.Send(query);
 
+            return Ok(new { product });
```
Comment (NdubuisiJr) [inline src/Hng.Web/Controllers/ProductController.cs:36]: Is this the correct response in the ticket?

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

**Review by NdubuisiJr — APPROVED** [review]

**Pull request merged**